snow/engine/snowman/transitive.go

```diff
@@ -454,15 +468,69 @@ func (t *Transitive) GetBlock(ctx context.Context, blkID ids.ID) (snowman.Block,
 	return t.VM.GetBlock(ctx, blkID)
 }
 
-func (t *Transitive) sendChits(ctx context.Context, nodeID ids.NodeID, requestID uint32) {
-	lastAccepted, _ := t.Consensus.LastAccepted()
+func (t *Transitive) sendChits(ctx context.Context, nodeID ids.NodeID, requestID uint32, requestedHeight uint64) {
+	lastAcceptedID, lastAcceptedHeight := t.Consensus.LastAccepted()
 	// If we aren't fully verifying blocks, only vote for blocks that are widely
 	// preferred by the validator set.
 	if t.Ctx.StateSyncing.Get() || t.Config.PartialSync {
-		t.Sender.SendChits(ctx, nodeID, requestID, lastAccepted, lastAccepted)
+		acceptedAtHeight, err := t.VM.GetBlockIDAtHeight(ctx, requestedHeight)
+		if err != nil {
+			// Because we only return accepted state here, it's fairly likely
+			// that the requested height is higher than the last accepted block.
+			// That means that this code path is actually quite common.
+			t.Ctx.Log.Debug("failed fetching accepted block",
+				zap.Stringer("nodeID", nodeID),
+				zap.Uint64("requestedHeight", requestedHeight),
+				zap.Uint64("lastAcceptedHeight", lastAcceptedHeight),
+				zap.Stringer("lastAcceptedID", lastAcceptedID),
+				zap.Error(err),
+			)
+			acceptedAtHeight = lastAcceptedID
```

Comment:
What are the implications of responding with an ID that is not actually the accepted ID at that height?  My understanding is that this occurs when the current node does not have any accepted id at height because it has not sync'd to that height yet.  Would it make more sense to return some sort of empty ID or does the response effectively not matter in this case?

Reply:
>height because it has not sync'd to that height yet.

this may also occur if the height is pruned or it was never fetched (bc state synced)

Reply:
Generally people should only be requesting heights "close" to tip here. This case specifically handles the case that it is _only_ safe to return Accepted state. It would also be safe to simply drop the requests here. But we do this to avoid causing the peer to hit a timeout.

It's safe for us to return any Accepted blocks here - but the most "useful" blocks to send are the lastAcceptedID + acceptedIDAtRequestedHeight. If we don't have access to the acceptedID at the requested height, we can return any accepted ID - which we just default to the last accepted ID.

Reply:
We could probably reduce our bandwidth a bit here by changing the proto format to have some logic for "inheriting" fields. In this code path we return 2-3 32-byte hashes but we could get away with sending 1-2 32-bytes hashes.

Reply:
Below - it is safe to return any preferred value.